src/test/hardfork/archive_migration_tests/steps.ml

```diff
@@ -0,0 +1,599 @@
+    | 0 ->
+        Deferred.unit
+    | _ ->
+        failwithf
+          "Expected to have at only canonical block while having %d orphaned \
+           and %d pending"
+          orphaned_blocks pending_blocks ()
+
+  let assert_migrated_db_contains_orphaned_and_pending_blocks query_migrated_db
+      =
+    let%bind pending_blocks =
+      query_migrated_db ~f:(fun db -> Sql.Berkeley.count_orphaned_blocks db)
+    in
+    let%bind orphaned_blocks =
+      query_migrated_db ~f:(fun db -> Sql.Berkeley.count_pending_blocks db)
+    in
+    match orphaned_blocks + pending_blocks with
+    | 0 ->
+        failwith "Expected to have at least one pending or orphaned block"
+    | _ ->
+        Deferred.unit
+
+  let compare_hashes mainnet_archive_conn_str migrated_archive_conn_str
+      end_global_slot ~full_migration =
+    let mainnet_archive_uri = Uri.of_string mainnet_archive_conn_str in
+    let migrated_archive_uri = Uri.of_string migrated_archive_conn_str in
+    let mainnet_pool =
+      Caqti_async.connect_pool ~max_size:128 mainnet_archive_uri
+    in
+    let migrated_pool =
+      Caqti_async.connect_pool ~max_size:128 migrated_archive_uri
+    in
+    match (mainnet_pool, migrated_pool) with
+    | Error _e, _ | _, Error _e ->
+        failwith "Failed to create Caqti pools for Postgresql"
+    | Ok mainnet_pool, Ok migrated_pool ->
+        let query_mainnet_db = Mina_caqti.query mainnet_pool in
+        let query_migrated_db = Mina_caqti.query migrated_pool in
+        let compare_hashes ~fetch_data_sql ~find_element_sql name =
+          let%bind expected_hashes = query_mainnet_db ~f:fetch_data_sql in
+          Deferred.List.iter expected_hashes ~f:(fun hash ->
+              let%bind element_id = find_element_sql hash in
+              if element_id |> Option.is_none then
+                failwithf "Cannot find %s hash ('%s') in migrated database" name
+                  hash ()
+              else Deferred.unit )
+        in
+        let%bind _ =
+          compare_hashes
+            ~fetch_data_sql:(fun db ->
+              Sql.Mainnet.user_commands_hashes db end_global_slot )
+            ~find_element_sql:(fun hash ->
+              query_migrated_db ~f:(fun db ->
+                  Sql.Berkeley.find_user_command_id_by_hash db hash ) )
+            "user_commands"
+        in
+        let%bind _ =
+          compare_hashes
+            ~fetch_data_sql:(fun db ->
+              Sql.Mainnet.internal_commands_hashes db end_global_slot )
+            ~find_element_sql:(fun hash ->
+              query_migrated_db ~f:(fun db ->
+                  Sql.Berkeley.find_internal_command_id_by_hash db hash ) )
+            "internal_commands"
+        in
+        let%bind _ =
+          match full_migration with
+          | true ->
+              assert_migrated_db_contains_only_canonical_blocks
+                query_migrated_db
+          | false ->
+              assert_migrated_db_contains_orphaned_and_pending_blocks
```

Comment:
This check doesn't quite make sense. If it's not full migration, let's say we just migrated the first 200 slots. Then those 200 blocks should all be canonical which makes this test to fail. We won't migrate orphaned blocks in `berkeley_migration` app. And there won't be any pending blocks too because mainnet blocks created within the first 200 slots should either be canonical or orphaned. This check would only pass if the mainnet database is corrupted somehow, that it has some blocks within the 200 slots that marked as pending

Reply:
Agree! I will fix the condition and assertion